samples/bluetooth/broadcast_audio_source/src/main.c

```diff
@@ -116,14 +239,92 @@ static void init_lc3(void)
 		return;
 	}
 
+#if !defined(CONFIG_USB_DEVICE_AUDIO)
+	/* If USB is not used as a sound source, generate a sine wave */
+	fill_audio_buf_sin(send_pcm_data, frame_duration_us, AUDIO_TONE_FREQUENCY_HZ, freq_hz);
+#endif
+
 	/* Create the encoder instance. This shall complete before stream_started() is called. */
-	lc3_encoder = lc3_setup_encoder(frame_duration_us, freq_hz, 0, /* No resampling */
-					&lc3_encoder_mem);
+	for (size_t i = 0U; i < ARRAY_SIZE(streams); i++) {
+		streams[i].lc3_encoder = lc3_setup_encoder(frame_duration_us,
+			freq_hz, 0, &streams[i].lc3_encoder_mem);
+
+		if (streams[i].lc3_encoder == NULL) {
+			printk("ERROR: Failed to setup LC3 encoder - wrong parameters?\n");
+		}
+	}
+
+	while (true) {
+		for (size_t i = 0U; i < ARRAY_SIZE(streams); i++) {
+			k_sem_take(&lc3_encoder_sem, K_FOREVER);
+		}
+		for (size_t i = 0U; i < ARRAY_SIZE(streams); i++) {
+			send_data(&streams[i]);
+		}
+	}
+}
 
-	if (lc3_encoder == NULL) {
-		printk("ERROR: Failed to setup LC3 encoder - wrong parameters?\n");
+#define LC3_ENCODER_STACK_SIZE	4096
+#define LC3_ENCODER_PRIORITY	5
+
+K_THREAD_DEFINE(encoder_tid, LC3_ENCODER_STACK_SIZE, init_lc3_thread,
+		NULL, NULL, NULL, LC3_ENCODER_PRIORITY, 0, -1);
+
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+static void data_received(const struct device *dev,
+			  struct net_buf *buffer,
+			  size_t size)
+{
+	int ret;
+	static int count;
+	int16_t *pcm;
+	int nsamples, ratio;
+	int16_t usb_pcm_data[2][USB_NUM_SAMPLES];
+
+
+	if (!buffer || !size) {
+		/* This should never happen */
+		return;
 	}
+
+	pcm = (int16_t *)net_buf_pull_mem(buffer, size);
+
+	/* 'size' is in bytes, containing 1ms, 48kHz, stereo, 2 bytes per sample.
+	 * Take left channel and do a simple downsample to 16kHz
+	 */
+
+	ratio = USB_SAMPLE_RATE / USB_DOWNSAMPLE_RATE;
+	nsamples = size / (sizeof(int16_t) * 2 * ratio);
+	for (size_t i = 0, j = 0; i < nsamples; i++, j += ratio*USB_CHANNELS) {
+		usb_pcm_data[0][i] = pcm[j];
+		usb_pcm_data[1][i] = pcm[j+1];
+	}
+
+	for (size_t i = 0U; i < MAX(ARRAY_SIZE(streams), 2); i++) {
+		ret = ring_buf_put(&(streams[i].audio_ring_buf),
+			(uint8_t *)(usb_pcm_data[i]), nsamples * 2);
+		if (ret < nsamples * 2) {
```

Comment:
```suggestion
	for (size_t i = 0U; i < MAX(ARRAY_SIZE(streams), USB_CHANNELS); i++) {
		const uint32_t total_nsamples = nsamples * USB_CHANNELS;
		
		ret = ring_buf_put(&(streams[i].audio_ring_buf),
			(uint8_t *)(usb_pcm_data[i]), total_nsamples);
		if (ret < total_nsamples) {
```

Reply:
this is multiplied by 2 because of 2 bytes per sample - not channels. But will see if I can make it more clear